cmd/snap-confine/mount-support.c

```diff
@@ -135,76 +138,51 @@ static void setup_private_pts()
 	sc_do_mount("/dev/pts/ptmx", "/dev/ptmx", "none", MS_BIND, 0);
 }
 
-/*
- * Setup mount profiles as described by snapd.
- *
- * This function reads /var/lib/snapd/mount/$security_tag.fstab as a fstab(5) file
- * and executes the mount requests described there.
- *
- * Currently only bind mounts are allowed. All bind mounts are read only by
- * default though the `rw` flag can be used.
+/**
+ * Setup mount profiles by running snap-update-ns.
  *
- * This function is called with the rootfs being "consistent" so that it is
- * either the core snap on an all-snap system or the core snap + punched holes
- * on a classic system.
+ * The first argument is an open file descriptor (though opened with O_PATH, so
+ * not as powerful), to a copy of snap-update-ns. The program is opened before
+ * the root filesystem is pivoted so that it is easier to pick the right copy.
  **/
-static void sc_setup_mount_profiles(const char *snap_name)
+static void sc_setup_mount_profiles(int snap_update_ns_fd,
+				    const char *snap_name)
 {
-	debug("%s: %s", __FUNCTION__, snap_name);
-
-	FILE *desired __attribute__ ((cleanup(sc_cleanup_endmntent))) = NULL;
-	FILE *current __attribute__ ((cleanup(sc_cleanup_endmntent))) = NULL;
-	char profile_path[PATH_MAX];
-
-	sc_must_snprintf(profile_path, sizeof(profile_path),
-			 "/run/snapd/ns/snap.%s.fstab", snap_name);
-	debug("opening current mount profile %s", profile_path);
-	current = setmntent(profile_path, "w");
-	if (current == NULL) {
-		die("cannot open current mount profile: %s", profile_path);
+	debug("calling snap-update-ns to initialize mount namespace");
+	pid_t child = fork();
+	if (child < 0) {
+		die("cannot fork to run snap-update-ns");
 	}
-
-	sc_must_snprintf(profile_path, sizeof(profile_path),
-			 "/var/lib/snapd/mount/snap.%s.fstab", snap_name);
-	debug("opening desired mount profile %s", profile_path);
-	desired = setmntent(profile_path, "r");
-	if (desired == NULL && errno == ENOENT) {
-		// It is ok for the desired profile to not exist. Note that in this
-		// case we also "update" the current profile as we already opened and
-		// truncated it above.
-		return;
+	if (child == 0) {
+		// We are the child, execute snap-update-ns
+		char *snap_name_copy
+		    __attribute__ ((cleanup(sc_cleanup_string))) = NULL;
+		snap_name_copy = strdup(snap_name);
+		if (snap_name_copy == NULL) {
+			die("cannot copy snap name");
+		}
+		char *argv[] = {
+			"snap-update-ns", "--from-snap-confine", snap_name_copy,
+			NULL
+		};
+		char *envp[] = { NULL };
+		debug("fexecv(%d (snap-update-ns), %s %s %s,)",
+		      snap_update_ns_fd, argv[0], argv[1], argv[2]);
+		fexecve(snap_update_ns_fd, argv, envp);
+		die("cannot execute snap-update-ns");
 	}
-	if (desired == NULL) {
-		die("cannot open desired mount profile: %s", profile_path);
+	// Wait for snap-update-ns to finish.
```

Comment:
`// We are the parent, so wait for snap-update-ns to finish`

Reply:
Done